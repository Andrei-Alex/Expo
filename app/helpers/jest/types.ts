import { ReactTestRendererJSON } from 'react-test-renderer';

export type ReactTreeType = ReactTestRendererJSON | null;
