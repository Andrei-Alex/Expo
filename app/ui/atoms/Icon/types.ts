import { OpaqueColorValue, ViewStyle } from 'react-native';

export type icons =
  | 'format-list-bulleted'
  | 'email'
  | 'logout'
  | 'home'
  | 'chevron-down'
  | 'apps'
  | 'lock'
  | 'camera'
  | 'account'
  | 'chevron-right';

export interface IIcon {
  name: icons;
  size?: number;
  backgroundColor?: string;
  iconColor?: string | OpaqueColorValue;
  testID?: string;
  style?: ViewStyle;
}
