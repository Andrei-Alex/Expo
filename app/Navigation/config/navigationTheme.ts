import { DefaultTheme } from '@react-navigation/native';
import { colors } from '../../config';

export const appTheme = {
  ...DefaultTheme,
  colors: {
    ...DefaultTheme.colors,
    primary: colors.primary,
    background: colors.white
  }
};
