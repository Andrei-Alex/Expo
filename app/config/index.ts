export { default as colors } from './theme/colors';
