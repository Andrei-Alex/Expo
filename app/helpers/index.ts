export * from './jest/jest';
export * from './jest/types';
export * from './mocks/navigationMock';
