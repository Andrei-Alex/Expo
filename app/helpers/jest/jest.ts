import { ReactTestRendererJSON } from 'react-test-renderer';
import { ReactTreeType } from './types';
/**
 * Recursively searches the rendered component tree for an element with a given testID.
 * @param {object} tree - The rendered component tree.
 * @param {string} testID - The testID to search for.
 * @returns {object|null} The found element, or null if not found.
 */

export function findByTestId(
  tree: ReactTreeType,
  testID: string
): ReactTreeType {
  if (!tree || (tree.props && tree.props.testID === testID)) {
    return tree;
  }
  if (tree.children) {
    for (let child of tree.children) {
      if (typeof child === 'object') {
        const found = findByTestId(child, testID);
        if (found) return found;
      }
    }
  }
  return null;
}
