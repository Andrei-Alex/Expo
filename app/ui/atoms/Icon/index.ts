export { styles } from './styles';
export * from './types';
