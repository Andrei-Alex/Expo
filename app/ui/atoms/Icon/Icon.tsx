import React from 'react';
import { View } from 'react-native';

import { styles, Props } from '.';
import { colors } from '../../../config';

import { MaterialCommunityIcons } from '@expo/vector-icons';

/**
 * Icon Component
 *
 * A reusable and customizable icon component that integrates MaterialCommunityIcons from '@expo/vector-icons' library, providing
 * a wide range of icons suitable for various application needs. It leverages the flexibility of React Native's View component
 * to create a container for the icon, which can be customized in size, background color, and icon color. This component
 * simplifies the process of adding icons to your app, ensuring consistency in design and ease of use.
 *
 * ### Usage
 * ```jsx
 * <Icon
 *   name="email"
 *   size={50}
 *   backgroundColor="#f0f"
 *   iconColor="#fff"
 * />
 * ```
 *
 * @component
 * @param {Object} props - The properties passed to the component.
 * @param {string} props.name - The name of the icon to be displayed.
 * @param {number} [props.size=40] - The size of the icon and its container. Defaults to 40.
 * @param {string} [props.backgroundColor=colors.black] - The background color of the icon container. Defaults to black.
 * @param {string} [props.iconColor=colors.white] - The color of the icon. Defaults to white.
 * @returns {React.ReactElement} A customizable icon component that can be easily integrated into React Native applications.
 */
const Icon: React.FC<Props> = ({
  name,
  size = 40,
  backgroundColor = colors.black,
  iconColor = colors.white
}) => {
  return (
    <View
      style={[
        styles.icon,
        {
          width: size,
          height: size,
          borderRadius: size / 2,
          backgroundColor
        }
      ]}
    >
      <MaterialCommunityIcons name={name} color={iconColor} size={size * 0.5} />
    </View>
  );
};

export default Icon;
