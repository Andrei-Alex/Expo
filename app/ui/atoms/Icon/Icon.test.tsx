import React from 'react';
import renderer from 'react-test-renderer';

import { colors } from '../../../config';
import Icon from './Icon';
import { findByTestId } from '../../../helpers/jest';

describe('Icon', () => {
  const defaultIconProps = {
    name: 'email',
    size: 50,
    backgroundColor: '#f0f',
    iconColor: '#fff',
    testID: 'email-icon'
  };
  it('renders correctly with default props', () => {
    const tree = renderer.create(<Icon {...defaultIconProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with custom size, backgroundColor, and iconColor', () => {
    const tree = renderer.create(<Icon {...defaultIconProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders multiple icons with distinct testIDs', () => {
    const component = renderer.create(
      <>
        <Icon {...defaultIconProps} />
        <Icon
          name="home"
          size={40}
          backgroundColor={colors.white}
          iconColor={colors.black}
          testID={'home-icon'}
        />
      </>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly and can be found by testID', () => {
    const component = renderer.create(<Icon {...defaultIconProps} />);

    const tree = component.toJSON();
    const iconContainer = findByTestId(tree, 'container-email-icon');

    expect(iconContainer).not.toBeNull();
    expect(iconContainer.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          backgroundColor: '#f0f'
        })
      ])
    );
  });
});
