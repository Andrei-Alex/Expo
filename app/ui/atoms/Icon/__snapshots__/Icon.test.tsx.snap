// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders correctly with custom size, backgroundColor, and iconColor 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#f0f",
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      undefined,
    ]
  }
  testID="container-email-icon"
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#fff",
          "fontSize": 25,
        },
        undefined,
        {
          "fontFamily": "material-community",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
    testID="icon-email-icon"
  >
    󰇮
  </Text>
</View>
`;

exports[`Icon renders correctly with default props 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#f0f",
        "borderRadius": 25,
        "height": 50,
        "width": 50,
      },
      undefined,
    ]
  }
  testID="container-email-icon"
>
  <Text />
</View>
`;

exports[`Icon renders multiple icons with distinct testIDs 1`] = `
[
  <View
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        {
          "backgroundColor": "#f0f",
          "borderRadius": 25,
          "height": 50,
          "width": 50,
        },
        undefined,
      ]
    }
    testID="container-email-icon"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#fff",
            "fontSize": 25,
          },
          undefined,
          {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="icon-email-icon"
    >
      󰇮
    </Text>
  </View>,
  <View
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        {
          "backgroundColor": "#fff",
          "borderRadius": 20,
          "height": 40,
          "width": 40,
        },
        undefined,
      ]
    }
    testID="container-home-icon"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#000",
            "fontSize": 20,
          },
          undefined,
          {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="icon-home-icon"
    >
      󰋜
    </Text>
  </View>,
]
`;
